### ISSUE TITLE
Update Run IDL Script (kill - related)

### ISSUE THREAD

**Issue by pcampanella** [fadeoutsoftware/WASDI#82]
In the IDL Docker the runProcessor.sh file is contained in the template.
This is the script file that the local WasdiProcessorServer starts to run a IDL processor.

The bash script run another process, that is the real ENVI/IDL processor.
It has been improved to be more safe: first of all sometimes randomly the start of IDL fails so we added a sort of retry. 
The line:
umask 000; /usr/local/bin/idl ${script_file}
is the one that run IDL. 
If after this line the status is still RUNNING, something bad is happening so the script makes up to 5 retry.

After these 5 tentatives, if the status is not one of the safe close one (DONE, ERROR, STOPPED) the script force an update status as ERROR.
This should work and is to be tested (already assigned to another issue).

We need to futher improve the bash script to have these new feature:
.Read and communicate in some way the PID of the real ENVI processor:
in the line umask 000; /usr/local/bin/idl ${script_file} is there we start idl: is possible to have that PID? This should be needed to implement the kill in IDL that, at the moment, does not have the docker pid to kill.

If all this works, the script should call the 
        
        WorkspaceBaseUrl + '/process/setsubpid'
with payload = {'procws': sProcessId, 'subpid': iSubPid}
to update the pid in the db.
If killing the .sh kills also idl, we should instead put the script pid in the db.

**Comment by wasdi-sysadmin-br** [comment]
**Update**

**Long version**

I worked on this problem with other problems on IDL processors.
We had 3 problems.

Problem 1/3:
When we launched edrift_flood_event, many instances of edriftlistflood_archive was executed.
When many instances edriftlistflood_archive was started:
  - one of them worked properly (the instance finished, we had logs, etc)
  - others instances failed without logs / without information
So the questions was: why it failed and how to fix?

Problem 2/3:
As just mentionned edriftlistflood_archive failed without log.
So the questions was: why no log and how to fix?

Problem 3/3:
To retrieve the PID of the IDL script instead of the bash's script PID.

Solution 1/3:
When an instance of edriftlistflood_archive is called, wasdiProcessorServer.py produces / updates a file param.properties with values.
If many instances are executed in a very short time, the file param.properties is updated by each instance.
The last instance will work and other will failed because parameters are wrong.

I proposed to Paolo to generate a file PER instance based on the processor ID but he has to test how to pass the argument to IDL.
So I'm waiting.

Solution 2/3:
The docker container edriftlistflood_archive was not started with the script runwasdidocker.sh but it seems it was started manually: so the logs were not redirected to the dockers.log file.
I stopped the docker container
I restarted it using the runwasdidocker.sh script
I fixed the IDL license server address
And now it works.

Solution 3/3:
My solution to retrieve the IDL script PID is:
  - I improved the script runProcessor.sh:
    - I added a command line argument manager
    - each IDL script instance is started in background with the character & at the end of the command line
    - the PID is getted with a special Bash variable: ${!}
    - the PID is saved in an unique file (the file path is given on the command line when wasdiProcessorServer.py call runProcessor.sh): file = /tmp/pid_idl_<processor ID>
  - I improved the Python server wasdiProcessorServer.py:
    - it calls/executes runProcessor.sh with parameters/arguments like the path to the PID file to produce
    - as IDL scripts are executed in background, runProcessor.sh is not a blocking point: the Python execution continues quickly
    - then I execute a new script getPid.sh to read the file /tmp/pid_idl_<processor ID>
    ---> we can read directly in Python but I prefered the getPid.sh approach because I added the possibility to say "hey please wait the PID file appears" and I use it in the Python code (we wait 300 seconds per pid file)
  - I wrote a new script getPid.sh:
    - this script just read /tmp/pid_idl_<processor ID>
    - we can decide to wait the PID file or not
    - if we call it with the argument to wait the PID file, we can configure a timeout

All what I did is available on PROD > ADWAISEO > 01 under /data/wasdi/processors/br_edriftlistflood_archive

---

**Short version**

For the problem with edriftlistflood_archive that failed when many instances are executed:
  - I suppose it's because wasdiProcessorServer.py  overwrite param.properties with values of the last instance
  - I suggested to Paolo to produce a param.properties specific per processor id (example: param.properties.<processor ID>) and to pass the name to the IDL script
  - Paolo said me he will check how to pass the value as argument when he will have time so I'm waiting
  
After (when Paolo will pass the name as argument) we can continue to finish the ticket.
What I did is available on PROD > ADWAISEO > 01 under /data/wasdi/processors/br_edriftlistflood_archive

**Comment by wasdi-sysadmin-br** [comment]
**Update**

We worked with Paolo yesterday to finalize what I did but above all to pass the value to the IDL script.
A first test yesterday evening seems to have worked.

This morning I just redeployed on adwaiseo 01 to update runProcessor.sh to remove a hardcoded value in logs + to implement the cleaning of params and config files.
A test is ongoing.

**Comment by wasdi-sysadmin-br** [comment]
**Update**

For me the last test is conclusive: I'm waiting Paolo's confirmation.

Files (params / config / pid) was cleaned
WASDI retrieved the right PID and so stored the right value in the DB
WASDI can kill an IDL script: it is marked as STOPPED in WASDI
All instances of edriftlistflood_archive was executed properly (when previously only one worked)
